### ISSUE TITLE
[Android]  Use drawables from resources opposed to creating bitmaps in order to improve performance

### ISSUE THREAD

**Issue by PureWeen** [xamarin/Xamarin.Forms#1908]
### Description
@jonathanpeppers outlines some issues here with how images are loaded with Android

http://jonathanpeppers.com/Blog/xamarin-forms-performance-on-android

> Notice how many Android.Graphics.Bitmaps get created here... If two rows in a ListView use the same image, they each use a copy of the exact same image. If you scroll a cell off screen and bring it back, it will load a new Bitmap object when it is brought back onto the screen. 

On android if we use a Drawable when present instead of the bitmap this helps with reuse of the same image.

Also looking into caching the Resource Ids against the names might be useful as the ResourceManager currently uses reflection

### Steps to Reproduce
Create a large number of images all based on the same Drawable in Android
```
StackLayout listView = new StackLayout();
             
            for (int i = 0; i < 1000; i++)
            {
                listView.Children.Add(new Image(){ Source = "oasis.jpg"});                
            }

            return new ContentPage()
            {
                Content = new ScrollView(){ Content = listView}
            };
```
Currently this will crash with an out of memory exception. 

### Expected Behavior
Use drawables where you can instead of constructing  bitmaps so listviews can handle lots of repetative images.

### Actual Behavior
Crashes with out of memory when using too many images

### Implementation

I've worked up a PoC here
https://github.com/xamarin/Xamarin.Forms/tree/ImagePerf

### Backwards Compatibility
Does this need to be a platform opt in? or will it work just dropped in. What dependencies are there if the ImageView has always been set via a bitmap

If the user isn't using Drawable does the extra check for a Drawable cause them a performance decrease for the lookup of the ID? 

Can the names inside ResourceManager all just be cached against ids? 

Reflection performance vs 
https://developer.android.com/reference/android/content/res/Resources.html#getIdentifier(java.lang.String,%20java.lang.String,%20java.lang.String)

**Comment by jonathanpeppers** [comment]
I think using `Drawable` is a good idea here, but I'm wondering if there is a different way to implement this.

`ImageViewExtensions` is `internal`... Does that mean that third party components or developer's custom renderers can't take advantage of this change? It makes me think that breaking changes are required here to make the best improvements...

So some ideas here:
1. This PR might be good in the interim since there are no breaking changes
2. `IImageSourceHandler`'s API could be changed to return a `Drawable`. `BitmapDrawable` could be used for all the places `Bitmap` is required.
3. Google's answer to `Bitmap` troubles is the Glide library as mentioned [here](https://developer.android.com/topic/performance/graphics/manage-memory.html). Could Xamarin.Forms or a separate NuGet fully take advantage of Glide? It seems to cover more than just images.

No. 3 could be the holy grail for fast images in Xamarin.Forms. Unforunately, it also seems like it would take the most research and work for it to pay off:
- We would need an official Xamarin NuGet for Glide
- XF would either need to take a dependency on Glide, or we make sure its possible for a separate library to modify the internals enough to create something like `Xamarin.Forms.Glide`

Just ideas here, let me know what everyone else thinks!

**Comment by PureWeen** [comment]
> ImageViewExtensions is internal

  Yea I could definitely see it being useful to split up the logic in *UpdateBitmap* method and providing some useful extension that's like   *FormsImage.ToAndroidDrawable() * 

> IImageSourceHandler's API could be changed to return a Drawable

This is what I was thinking would be ideal as well but I'll have to see the viability of doing that :-)  I've asked internally about it. 

>Google's answer to Bitmap troubles is the Glide 

I found this bindings project which has some decent use 
https://github.com/beraybentesen/glide-xamarin-android

If Xamarin created a binding project for glide I'm thinking it'd be part of the *Xamarin.Android.Glide* namespace and since ```Xamarin.Forms.Glide``` would insinuate something that's going to give you the same behaviors on UWP and iOS as well

But an official Glide binding project would be pretty neat!!

> but I'm wondering if there is a different way to implement this

Same, but I haven't quite hit on anything for maintaining backwards compatibility that I like much more yet...

I was toying with something like this
```
[Obsolete]
public interface IImageSourceHandler : IRegisterable
	{
		Task<Bitmap> LoadImageAsync(ImageSource imagesource, Context context, CancellationToken cancelationToken = default(CancellationToken));
	}

	public interface IImageDrawableSourceHandler : IImageSourceHandler
	{
		Task<Drawable> LoadDrawableAsync(ImageSource imagesource, Context context, CancellationToken cancelationToken = default(CancellationToken));
	}
```

But that makes the implementation of *IImageDrawableSourceHandler* awkward for users (are default interface methods a thing yet?)  ..

**Comment by jonathanpeppers** [comment]
@PureWeen I have a prototype using Glide: https://github.com/jonathanpeppers/glidex

It seems to perform very well so far!

**Comment by PureWeen** [comment]
That's awesome @jonathanpeppers !! Everyone on the forms team got really excited by this :-) I've pulled it into a couple of my pet projects.   It's great how it just seamlessly takes over with the init

Also we're going to get the small fixed I made merged in for 3.0.0
https://github.com/xamarin/Xamarin.Forms/pull/1973

Which should help on the raw forms side of things

Piece by piece fixing up the images sore spots!!
